File: grx249/test/polytest.c
/**
 ** polytest.c ---- test polygon rendering
 **
 ** Copyright (c) 1995 Csaba Biegl, 820 Stirrup Dr, Nashville, TN 37221
 ** [e-mail: csaba@vuse.vanderbilt.edu]
 **
 ** This is a test/demo file of the GRX graphics library.
 ** You can use GRX test/demo files as you want.
 **
 ** The GRX graphics library is free software; you can redistribute it
 ** and/or modify it under some conditions; see the "copying.grx" file
 ** for details.
 **
 ** This library is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 **
 **/

#include <string.h>
#include <stdio.h>

#include "test.h"

static GrColor *EGA;
#define black EGA[0]
#define red   EGA[12]
#define blue  EGA[1]
#define white EGA[15]

static void testpoly(int n,int points[][2],int convex)
{
	GrClearScreen(black);
	GrPolygon(n,points,white);
	GrFilledPolygon(n,points,(red | GrXOR));
	GrKeyRead();
	if(convex || (n <= 3)) {
	    GrClearScreen(black);
	    GrFilledPolygon(n,points,white);
	    GrFilledConvexPolygon(n,points,(red | GrXOR));
	    GrKeyRead();
	}
}

static void speedtest(void)
{
	int pts[4][2];
	int ww = GrSizeX() / 10;
	int hh = GrSizeY() / 10;
	int sx = (GrSizeX() - 2*ww) / 32;
	int sy = (GrSizeY() - 2*hh) / 32;
	int  ii,jj;
	GrColor color;
	long t1,t2,t3;
   char msg[81];

	GrClearScreen(black);
	t1 = GrMsecTime();
	pts[0][1] = 0;
	pts[1][1] = hh;
	pts[2][1] = 2*hh;
	pts[3][1] = hh;
	color = 0;
	for(ii = 0; ii < 32; ii++) {
	    pts[0][0] = ww;
	    pts[1][0] = 2*ww;
	    pts[2][0] = ww;
	    pts[3][0] = 0;
	    for(jj = 0; jj < 32; jj++) {
		GrFilledPolygon(4,pts, EGA[color] | GrXOR);
		color = (color + 1) & 15;
		pts[0][0] += sx;
		pts[1][0] += sx;
		pts[2][0] += sx;
		pts[3][0] += sx;
	    }
	    pts[0][1] += sy;
	    pts[1][1] += sy;
	    pts[2][1] += sy;
	    pts[3][1] += sy;
	}
	t2 = GrMsecTime();
	pts[0][1] = 0;
	pts[1][1] = hh;
	pts[2][1] = 2*hh;
	pts[3][1] = hh;
	color = 0;
	for(ii = 0; ii < 32; ii++) {
	    pts[0][0] = ww;
	    pts[1][0] = 2*ww;
	    pts[2][0] = ww;
	    pts[3][0] = 0;
	    for(jj = 0; jj < 32; jj++) {
		GrFilledConvexPolygon(4,pts, EGA[color] | GrXOR);
		color = (color + 1) & 15;
		pts[0][0] += sx;
		pts[1][0] += sx;
		pts[2][0] += sx;
		pts[3][0] += sx;
	    }
	    pts[0][1] += sy;
	    pts[1][1] += sy;
	    pts[2][1] += sy;
	    pts[3][1] += sy;
	}
	t3 = GrMsecTime();
	GrTextXY(0, 0, "Times to scan 1024 polygons", white, black);
	sprintf(msg, "   with 'GrFilledPolygon': %.2f (s)",
	       (double)(t2 - t1) / (double)1000);
	GrTextXY(0, 18, msg, white, black);
	sprintf(msg, "   with 'GrFilledConvexPolygon': %.2f (s)",
	       (double)(t3 - t2) / (double)1000);
	GrTextXY(0, 36, msg, white, black);
}

TESTFUNC(ptest)
{
	char buff[300];
	int  pts[300][2];
	int  ii,collect;
	int  convex;
	FILE *fp;

	fp = fopen("polytest.dat","r");
	if(fp == NULL) return;
	EGA = GrAllocEgaColors();
	ii  = collect = convex = 0;
	while(fgets(buff,299,fp) != NULL) {
	    if(!collect) {
		if(strncmp(buff,"begin",5) == 0) {
		    convex  = (buff[5] == 'c');
		    collect = 1;
		    ii      = 0;
		}
		continue;
	    }
	    if(strncmp(buff,"end",3) == 0) {
		testpoly(ii,pts,convex);
		collect = 0;
		continue;
	    }
	    if(sscanf(buff,"%d %d",&pts[ii][0],&pts[ii][1]) == 2) ii++;
	}
	fclose(fp);
	speedtest();
	GrKeyRead();
}

File: grx249/pascal/polytest.pas
(**
 ** polytest.pas ---- test polygon rendering
 **
 ** Copyright (c) 1995 Csaba Biegl, 820 Stirrup Dr, Nashville, TN 37221
 ** [e-mail: csaba@vuse.vanderbilt.edu]
 **
 ** This is a test/demo file of the GRX graphics library.
 ** You can use GRX test/demo files as you want.
 **
 ** The GRX graphics library is free software; you can redistribute it
 ** and/or modify it under some conditions; see the "copying.grx" file
 ** for details.
 **
 ** This library is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 **
 **)

{$R-}

program polytest;

uses GPC, GRX, Test;

type WrkString       = String[80];

var  f               : Text;
     Line,msg        : WrkString;
     nb              : Integer;
     convex,collect  : boolean;
     k               : GrKeyType;
     polygon         : array [0..300] of PointType;
     pEGA            : GrColorsPtr;
     black,white,red : GrColor;

procedure TestPoly(n:Integer; var Points: array of PointType; convex:boolean);
begin
   GrClearScreen(black);
   GrPolygon(n,Points,white);
   GrFilledPolygon(n,Points,(red or GrXor));
   k:=GrKeyRead;
   if convex or (n <= 3) then begin
      GrClearScreen(black);
      GrFilledPolygon(n,Points,white);
      GrFilledConvexPolygon(n,Points,(red or GrXor));
      k:=GrKeyRead
   end
end;

procedure SpeedTest;
var
   pts : array[0..3,0..1] of CInteger;
   ww  : Integer = GrSizeX div 10;
   hh  : Integer = GrSizeY div 10;
   sx  : Integer = (GrSizeX - 2*ww) div 32;
   sy  : Integer = (GrSizeY - 2*hh) div 32;
   ii,jj : Integer;
   color : GrColor;
   t1,t2,t3,mu1,mu2,mu3: CInteger;
begin
   GrClearScreen(black);
   t1 := GetCPUTime(mu1);
   pts[0][1] := 0;
   pts[1][1] := hh;
   pts[2][1] := 2*hh;
   pts[3][1] := hh;
   color := 0;
   for ii := 0 to 31 do begin
      pts[0][0] := ww;
      pts[1][0] := 2*ww;
      pts[2][0] := ww;
      pts[3][0] := 0;
      for jj := 0 to 31 do begin
         GrFilledPolygon(4,pts, pEGA^[color] or GrXor);
         color := (color + 1) and 15;
         Inc(pts[0][0],sx);
         Inc(pts[1][0],sx);
         Inc(pts[2][0],sx);
         Inc(pts[3][0],sx);
      end;
      Inc(pts[0][1],sy);
      Inc(pts[1][1],sy);
      Inc(pts[2][1],sy);
      Inc(pts[3][1],sy);
   end;
   t2 := GetCPUTime(mu2);
   pts[0][1] := 0;
   pts[1][1] := hh;
   pts[2][1] := 2*hh;
   pts[3][1] := hh;
   color := 0;
   for ii := 0 to 31 do begin
      pts[0][0] := ww;
      pts[1][0] := 2*ww;
      pts[2][0] := ww;
      pts[3][0] := 0;
      for jj := 0 to 31 do begin
         GrFilledConvexPolygon(4,pts, pEGA^[color] or GrXor);
         color := (color + 1) and 15;
         Inc(pts[0][0],sx);
         Inc(pts[1][0],sx);
         Inc(pts[2][0],sx);
         Inc(pts[3][0],sx);
      end;
      Inc(pts[0][1],sy);
      Inc(pts[1][1],sy);
      Inc(pts[2][1],sy);
      Inc(pts[3][1],sy);
   end;
   t3 := GetCPUTime(mu3);
   GrTextXY(0, 0, 'Times to scan 1024 polygons', white, black);
   WriteStr(msg, ' with GrFilledPolygon: ',((t2+mu2/1e6)-(t1+mu1/1e6)):0:2,' (s)');
   GrTextXY(0, 18, msg, white, black);
   WriteStr(msg, '  with GrFilledConvexPolygon: ',((t3+mu3/1e6)-(t2+mu2/1e6)):0:2,' (s)');
   GrTextXY(0, 36, msg, white, black);
end;

begin
   InitTest;
   pEGA:=GrAllocEgaColors;
   black:=pEGA^[0]; red:=pEGA^[12]; white:=pEGA^[15];

   Assign(f,'../test/polytest.dat'); Reset(f);
   collect:=false;
   while not eof(f) do begin
      ReadLn(f,Line);
      if not collect then begin
         if Copy(Line,1,5)='begin' then begin
            collect:=true;
            convex := Line[6]='c';
            nb:=0;
         end
      end else begin
         if Copy(Line,1,3)='end' then begin
            if nb>0 then TestPoly(nb,polygon,convex);
            collect:=false;
         end else begin
            with polygon[nb] do ReadStr(Line,x, y);
            Inc(nb);
         end;
      end;
   end;
   Close(f);
   SpeedTest;
   k:=GrKeyRead;
   EndTest
end.
